File: plugins/nkn-sdk-flutter/ios/Classes/impl/Crypto.swift
import Nkn

class Crypto : ChannelBase, IChannelHandler, FlutterStreamHandler {
    
    let CHANNEL_NAME = "org.nkn.sdk/crypto"
    let EVENT_NAME = "org.nkn.sdk/crypto/event"
    var methodChannel: FlutterMethodChannel?
    var eventChannel: FlutterEventChannel?
    var eventSink: FlutterEventSink?
    
    func install(binaryMessenger: FlutterBinaryMessenger) {
        self.methodChannel = FlutterMethodChannel(name: CHANNEL_NAME, binaryMessenger: binaryMessenger)
        self.methodChannel?.setMethodCallHandler(handle)
    }
    
    func uninstall() {
        self.methodChannel?.setMethodCallHandler(nil)
    }
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
    
    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method{
        case "getPublicKeyFromPrivateKey":
            getPublicKeyFromPrivateKey(call, result: result)
        case "getPrivateKeyFromSeed":
            getPrivateKeyFromSeed(call, result: result)
        case "getSeedFromPrivateKey":
            getSeedFromPrivateKey(call, result: result)
        case "sign":
            sign(call, result: result)
        case "verify":
            verify(call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func getPublicKeyFromPrivateKey(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [String: Any]()
        let privateKey = args["privateKey"] as? FlutterStandardTypedData
        
        let publicKey = CryptoGetPublicKeyFromPrivateKey(privateKey?.data)
        result(publicKey)
    }
    
    private func getPrivateKeyFromSeed(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [String: Any]()
        let seed = args["seed"] as? FlutterStandardTypedData
        
        let privateKey = CryptoGetPrivateKeyFromSeed(seed?.data)
        result(privateKey)
    }
    
    private func getSeedFromPrivateKey(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [String: Any]()
        let privateKey = args["privateKey"] as? FlutterStandardTypedData
        
        let seed = CryptoGetSeedFromPrivateKey(privateKey?.data)
        result(seed)
    }
    
    private func sign(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [String: Any]()
        let privateKey = args["privateKey"] as? FlutterStandardTypedData
        let data = args["data"] as? FlutterStandardTypedData
        
        var error: NSError?
        let signature = CryptoSign(privateKey?.data, data?.data, &error)
        if (error != nil) {
            self.resultError(result: result, error: error)
            return
        }
        result(signature)
    }
    
    private func verify(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [String: Any]()
        let publicKey = args["publicKey"] as? FlutterStandardTypedData
        let data = args["data"] as? FlutterStandardTypedData
        let signature = args["signature"] as? FlutterStandardTypedData
        var error: NSError?
        
        CryptoVerify(publicKey?.data, data?.data, signature?.data, &error)
        if (error != nil) {
            //NSLog("%@", error!)
            result(false)
            return
        }
        result(true)
    }
}

File: plugins/nkn-sdk-flutter/android/src/main/kotlin/org/nkn/sdk/impl/Crypto.kt
package org.nkn.sdk.impl

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.nkn.sdk.IChannelHandler

class Crypto : IChannelHandler, MethodChannel.MethodCallHandler, EventChannel.StreamHandler, ViewModel() {
    companion object {
        val CHANNEL_NAME = "org.nkn.sdk/crypto"
        val EVENT_NAME = "org.nkn.sdk/crypto/event"
    }

    lateinit var channel: MethodChannel
    var eventSink: EventChannel.EventSink? = null

    override fun install(binaryMessenger: BinaryMessenger) {
        channel = MethodChannel(binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler(this)
    }

    override fun uninstall() {
        channel.setMethodCallHandler(null)
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getPublicKeyFromPrivateKey" -> {
                getPublicKeyFromPrivateKey(call, result)
            }
            "getPrivateKeyFromSeed" -> {
                getPrivateKeyFromSeed(call, result)
            }
            "getSeedFromPrivateKey" -> {
                getSeedFromPrivateKey(call, result)
            }
            "sign" -> {
                sign(call, result)
            }
            "verify" -> {
                // verify(call, result)
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    private fun getPublicKeyFromPrivateKey(call: MethodCall, result: MethodChannel.Result) {
        val privateKey = call.argument<ByteArray>("privateKey")
        val publicKey = crypto.Crypto.getPublicKeyFromPrivateKey(privateKey)
        result.success(publicKey)
    }

    private fun getPrivateKeyFromSeed(call: MethodCall, result: MethodChannel.Result) {
        val seed = call.argument<ByteArray>("seed")
        val privateKey = crypto.Crypto.getPrivateKeyFromSeed(seed)
        result.success(privateKey)
    }

    private fun getSeedFromPrivateKey(call: MethodCall, result: MethodChannel.Result) {
        val privateKey = call.argument<ByteArray>("privateKey")
        val seed = crypto.Crypto.getSeedFromPrivateKey(privateKey)
        result.success(seed)
    }

    private fun sign(call: MethodCall, result: MethodChannel.Result) {
        val privateKey = call.argument<ByteArray>("privateKey")
        val data = call.argument<ByteArray>("data")
        try {
            val signature = crypto.Crypto.sign(privateKey, data)
            result.success(signature)
        } catch (e: Throwable) {
            result.error("", e.localizedMessage, e.message)
        }
    }
}